src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -131,6 +141,55 @@ internal void Write(byte[] buffer, int offset, int count)
             ProcessWrite(buffer, offset, count, null);
         }
 
+        internal Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (!_sslState.InnerStream.CanWrite) return Task.FromException(new NotSupportedException(SR.NotSupported_UnwritableStream));
+            // If cancellation was requested, bail early with an already completed task.
+            // Otherwise, return a task that represents the Begin/End methods.
+            return cancellationToken.IsCancellationRequested
+                        ? Task.FromCanceled(cancellationToken)
+                        : WriteAsyncImpl(buffer, offset, count, cancellationToken);
+        }
+
+        private async Task WriteAsyncImpl(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            var semaphore = EnsureAsyncActiveWriteSemaphoreInitialized();
+            await semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+            InitiateWrite(buffer, offset, count, null);
+
+            bool failed = false;
+            try
+            {
+                await StartWritingAsync(buffer, offset, count, cancellationToken).ConfigureAwait(false);
+
+                if (Interlocked.Exchange(ref _nestedWrite, 0) == 0)
+                {
+                    throw new InvalidOperationException(SR.Format(SR.net_io_invalidendcall, "WriteAsyncImpl"));
+                }
+            }
+            catch (Exception e)
+            {
+                _sslState.FinishWrite();
+
+                failed = true;
+                if (e is IOException)
+                {
+                    throw;
+                }
+
+                throw new IOException(SR.net_io_write, e);
+            }
+            finally
+            {
+                semaphore.Release();
```

Comment:
Once the semaphore is removed, then this finally block can just go away, the failed bool can go away, and _nestedWrite can be set to 0 in the catch block.